### PR TITLE
core: a couple of follow-ups for issues found by fuzz-manager-serialize

### PR DESCRIPTION
**Pull request by mrc0mmand**

### CODE REVIEW

File: src/core/socket.c [hunk 1]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching fifo socket found: %s", value+skip);
+                        return 0;
                 }
 
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid fifo value: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: If `safe_atoi` fails, `fd` is uninitialized here (Use `fdv` for logging?)

File: src/core/socket.c [hunk 2]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching special socket found: %s", value+skip);
+                        return 0;
+                }
+
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid special value: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: same

File: src/core/socket.c [hunk 3]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching mqueue socket found: %s", value+skip);
+                        return 0;
+                }
+
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid mqueue value: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: same

File: src/core/socket.c [hunk 4]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                log_unit_debug(u, "No matching %s socket found: %s",
-                                               socket_address_type_to_string(type), value+skip);
+                r = extract_first_word(&value, &fdv, NULL, 0);
+                if (r <= 0) {
+                        log_unit_debug(u, "Failed to parse socket fd from value: %s", value);
+                        return 0;
+                }
+
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid socket fd: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: same here

File: src/core/socket.c [hunk 5]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
+                }
+
+                r = extract_first_word(&value, &typev, NULL, 0);
+                if (r <= 0) {
+                        log_unit_debug(u, "Failed to parse socket type from value: %s", value);
+                        return 0;
+                }
+
+                if (safe_atoi(typev, &type) < 0 || type < 0) {
+                        log_unit_debug(u, "Invalid socket type: %d", type);
```
Comment (YHNdnzj) [inline src/core/socket.c]: same

File: src/core/socket.c [hunk 6]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching netlink socket found: %s", value+skip);
+                        return 0;
                 }
 
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid socket value: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: same

File: src/core/socket.c [hunk 7]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching ffs socket found: %s", value+skip);
+                        return 0;
+                }
+
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
+                        log_unit_debug(u, "Invalid ffs value: %d", fd);
```
Comment (YHNdnzj) [inline src/core/socket.c]: last one :)

Comment (mrc0mmand) [reply src/core/socket.c]: Oh crap, that's what I get for copy-pasting :)

**Review by YHNdnzj — APPROVED** [review]

File: src/shared/fdset.c [hunk 8]
```diff
@@ -115,13 +119,23 @@ bool fdset_contains(FDSet *s, int fd) {
...
         return !!set_get(MAKE_SET(s), FD_TO_PTR(fd));
 }
 
 int fdset_remove(FDSet *s, int fd) {
         assert(s);
         assert(fd >= 0);
 
+        /* Avoid integer overflow in FD_TO_PTR() */
+        if (fd == INT_MAX)
+                log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "Refusing invalid fd: %d", fd);
```
Comment (bluca) [inline src/shared/fdset.c]: return ?

File: src/core/unit.c [hunk 9]
```diff
@@ -6162,7 +6162,7 @@ int activation_details_deserialize(const char *key, const char *value, Activatio
                         return -EINVAL;
 
                 t = unit_type_from_string(value);
-                if (t == _UNIT_TYPE_INVALID)
+                if (t == _UNIT_TYPE_INVALID || !IN_SET(t, UNIT_PATH, UNIT_TIMER))
```
Comment (bluca) [inline src/core/unit.c]: why are paths and timers singled out here?

Comment (bluca) [reply src/core/unit.c]: also the first check is redundant

Comment (bluca) [reply src/core/unit.c]: ie, if there's a specific reason, please add a comment as it's not obvious from the snippet

File: src/shared/fdset.c (hunk 8 above)
Comment (mrc0mmand) [reply src/shared/fdset.c]: :facepalm:

File: src/core/unit.c (hunk 9 above)
Comment (mrc0mmand) [reply src/core/unit.c]: The activation details vtable has ops defined only for paths and timers:

https://github.com/systemd/systemd/blob/6a85ca9bf513b0fff7d4385661e61c559f2a554e/src/core/unit.c#L6099-L6102

Comment (bluca) [reply src/core/unit.c]: got it - please add a comment above the check

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
I guess it'd make sense to wait for https://github.com/systemd/systemd/pull/27536 and then use it here as well instead of the first commit.

**Comment by bluca** [comment]
> I guess it'd make sense to wait for #27536 and then use it here as well instead of the first commit.

merged, needs a rebase

File: src/core/socket.c [hunk 10]
```diff
@@ -2651,124 +2652,177 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
-                                        p->fd = fdset_remove(fds, fd);
-                                        found = true;
-                                        break;
-                                }
-                        if (!found)
-                                log_unit_debug(u, "No matching fifo socket found: %s", value+skip);
+                        return 0;
                 }
 
+                if (safe_atoi(fdv, &fd) < 0 || fd < 0 || !fdset_contains(fds, fd)) {
```
Comment (bluca) [inline src/core/socket.c]: did you mean to switch this to parse_fd()?

Comment (mrc0mmand) [reply src/core/socket.c]: *sigh*, I did, but for some reason I forgot to commit it

**Review by bluca — APPROVED** [review]

**Pull request merged**